pkg/experiment/ingester/client/client.go

```diff
@@ -203,45 +202,53 @@ func (c *Client) stopping(_ error) error {
 func (c *Client) Push(
 	ctx context.Context,
 	req *segmentwriterv1.PushRequest,
-) (*segmentwriterv1.PushResponse, error) {
-	k := distributor.NewTenantServiceDatasetKey(req.TenantId, req.Labels)
-	p, dErr := c.distributor.Distribute(k, c.ring)
+) (resp *segmentwriterv1.PushResponse, err error) {
+	k := distributor.NewTenantServiceDatasetKey(req.TenantId, req.Labels...)
+	p, dErr := c.distributor.Distribute(k)
 	if dErr != nil {
 		_ = level.Error(c.logger).Log(
 			"msg", "unable to distribute request",
 			"tenant", req.TenantId,
-			"err", dErr)
+			"err", dErr,
+		)
 		return nil, status.Error(codes.Unavailable, errServiceUnavailableMsg)
 	}
 
+	// In case of a failure, the request is sent to another instance.
+	// At most 5 attempts to push the data to the segment writer.
+	instances := placement.ActiveInstances(p.Instances)
 	req.Shard = p.Shard
-	for { // The caller should cancel the context to break the loop.
-		instance, ok := p.Next()
-		if !ok {
-			_ = level.Error(c.logger).Log(
-				"msg", "no segment writer instances available for the request",
-				"tenant", req.TenantId)
-			return nil, status.Error(codes.Unavailable, errServiceUnavailableMsg)
-		}
-
-		resp, err := c.pushToInstance(ctx, req, instance.Addr)
-		// Happy path.
+	for attempts := 5; attempts >= 0 && instances.Next() && ctx.Err() == nil; attempts-- {
```

Comment:
Unlikely. Therefore, I hardcoded it here. I can't find a reason why anyone would need to change this.

The surrounding code for clarity:
```go
	// In case of a failure, the request is sent to another instance.
	// At most 5 attempts to push the data to the segment writer.
	instances := placement.ActiveInstances(p.Instances)
	req.Shard = p.Shard
	for attempts := 5; attempts >= 0 && instances.Next() && ctx.Err() == nil; attempts-- {
		instance := instances.At()
```

We have something very similar in the current version: https://github.com/grafana/pyroscope/blob/main/pkg/distributor/distributor.go#L523